File: leetcode/Code/221. 最大正方形.cpp
/**
* Author: Bao Wenjie
* Date: 2021/1/24
* Link: https://leetcode-cn.com/problems/maximal-square/
*/

#include <vector>
#include <iostream>
using namespace::std;


class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        int max_ = 0;
        vector<vector<int>> dp(m, vector<int>(n, 0));
        for (int j = 0; j < n; j++)
        {
            if (matrix[0][j] == '0')
                dp[0][j] = 0;
            else
                dp[0][j] = 1;
            max_ = max(dp[0][j], max_);
        }
        for (int i = 1; i < m; i++)
        {
            if (matrix[i][0] == '0')
                dp[i][0] = 0;
            else
                dp[i][0] = 1;
            max_ = max(dp[i][0], max_);
            for (int j = 1; j < n; j++)
            {
                if (matrix[i][j] == '0')
                    dp[i][j] = 0;
                else
                {
                    dp[i][j] = min(dp[i - 1][j - 1], min(dp[i - 1][j], dp[i][j - 1])) + 1;
                    max_ = max(dp[i][j], max_);
                }
            }
        }
        return max_ * max_;
    }
};

int main()
{
    vector<vector<char>> matrix = { {"1", "0", "1", "0", "0"},
                                    {"1", "0", "1", "1", "1"},
                                    {"1", "1", "1", "1", "1"},
                                    {"1", "0", "0", "1", "0"} };
    Solution so;
    so.maximalSquare(matrix);

}

File: leetcode/221. 最大正方形.py
# 在一个由 0 和 1 组成的二维矩阵内，找到只包含 1 的最大正方形，并返回其面积。

# 示例:

# 输入:

# 1 0 1 0 0
# 1 0 1 1 1
# 1 1 1 1 1
# 1 0 0 1 0

# 输出: 4


class Solution:
    def maximalSquare(self, matrix) -> int:
        maxSize = 0
        x = len(matrix)
        if(x == 0):
            return 0
        y = len(matrix[0])
        for k in range(y):
            if(matrix[0][k] == '1'):
                matrix[0][k] = 1
                maxSize = 1
            else:
                matrix[0][k] = 0
        for i in range(1, x):
            if(matrix[i][0] == '1'):
                matrix[i][0] = 1
                if(maxSize == 0):
                    maxSize = 1
            else:
                matrix[i][0] = 0
            for j in range(1, y):
                if(matrix[i][j] == '1'):
                    if(matrix[i - 1][j - 1] > 0 and
                       matrix[i - 1][j] > 0 and
                       matrix[i][j - 1] > 0):
                        matrix[i][j] = matrix[i - 1][j - 1] + 1
                    else:
                        matrix[i][j] = 0
                    if(matrix[i][j] > maxSize):
                            maxSize = matrix[i][j]
                else:
                    matrix[i][j] = 0
        print(matrix)
        return maxSize**2


matrix = [["0","0","0","1"],["1","1","0","1"],["1","1","1","1"],["0","1","1","1"],["0","1","1","1"]]
s = Solution()
print(s.maximalSquare(matrix))
